File: src/csharp/rehamove.cs
using UnityEngine;
using System;
using System.Linq;

namespace UnityRehamove {

	public class Rehamove {
		public RehamoveDevice rehamove;
		public int mode;

		public static string current_version = "v1.6";

		public static string[] channel0 = { "r", "red" };
		public static string[] channel1 = { "b", "blue" };
		public static string[] channel2 = { "g1", "gray1", "grey1", "black" };
		public static string[] channel3 = { "g2", "gray2", "grey2", "white" };

		public static int MODE_LOW_LEVEL = 0;
		public static int MODE_MID_LEVEL = 1;

		public Rehamove(string port_name) {
			rehamove = rehamovelib.open_port(port_name);
			mode = 0;
			if (rehamove == null) {
				Debug.Log("C# Rehamove init() ERROR!");
			} else {
				Debug.Log("C# Rehamove init() SUCCESS.");
			}
		}

		public string version() {
			string c_version = rehamovelib.get_version();
			Debug.Log("Rehamove Version: Python-side " + Rehamove.current_version.ToString() + ", C-side " + c_version);
			return Rehamove.current_version;
		}

		public int get_mode() {
			int result = rehamovelib.get_mode(rehamove);
			return result;			
		}

		public float get_current() {
			float result = rehamovelib.get_current(rehamove);
			return result;			
		}

		public int get_pulse_width() {
			int result = rehamovelib.get_pulse_width(rehamove);
			return result;			
		}

		public string info() {
			if (rehamove == null) {
				Debug.Log("C# Rehamove info() ERROR! Rehamove object does not exist.");
				return null;
			}
			int mode = get_mode();
			float current = get_current();
			int pulse_width = get_pulse_width();

			string output;
			if (mode == Rehamove.MODE_LOW_LEVEL) {
				output = String.Format("Rehamove device in low-level mode. Mid-level pulse is set to {0} mA and {1} us.", current, pulse_width);
			} else if (mode == Rehamove.MODE_MID_LEVEL) {
				output = String.Format("Rehamove device in mid-level mode. Mid-level pulse is set to {0} mA and {1} us.", current, pulse_width);
			} else {
				output = "Rehamove info() ERROR!";
			}

			return output;
		}

		public int get_channel(string channel) {
			int chosen_channel = 0; // Default
			if (channel0.Contains(channel)) {
				chosen_channel = 0;
			} else if (channel1.Contains(channel)) {
				chosen_channel = 1;
			} else if (channel2.Contains(channel)) {
				chosen_channel = 2;
			} else if (channel3.Contains(channel)) {
				chosen_channel = 3;
			}
			return chosen_channel;
		}

		public int get_channel(int channel) {
			if ((channel < 0) || (channel > 3)) {
				channel = 0; // Default
			}
			return channel;
		}
		
		public int pulse(string channel, float current, int pulse_width) {
			int converted_channel = get_channel(channel);
			return pulse(converted_channel, current, pulse_width);
		}

		public int pulse(int channel, float current, int pulse_width) {
			if (rehamove == null) {
				Debug.Log("C# Rehamove pulse() ERROR! Rehamove object does not exist.");
				return -1;
			}
			if (mode != MODE_LOW_LEVEL) {
				Debug.Log("C# Rehamove pulse() ERROR! Mode must be set to low-level.");
				return -1;
			}
			int result = rehamovelib.pulse(rehamove, channel, current, pulse_width);
			if (result != 0) {
				Debug.Log("C# Rehamove pulse() ERROR!");
				return -1;
			} else {
				Debug.Log("C# Rehamove pulse() sent.");
				return 0;
			}
		}

		public int battery() {
			if (rehamove == null) {
				Debug.Log("C# Rehamove battery() ERROR! Rehamove object does not exist.");
				return -1;
			}
			int result = rehamovelib.battery_request(rehamove);
			if (result != 0) {
				Debug.Log("C# Rehamove battery() ERROR!");
				return -1;
			} else {
				int battery_level = rehamovelib.get_battery(rehamove);
				Debug.Log("C# Rehamove battery(): " + battery_level.ToString() + "%");
				return battery_level;
			}
		}

		public int change_mode(int new_mode) {
			if (rehamove == null) {
				Debug.Log("C# Rehamove change_mode() ERROR! Rehamove object does not exist.");
				return -1;
			}
			int result = rehamovelib.change_mode(rehamove, new_mode);
			if (result != 0) {
				Debug.Log("C# Rehamove change_mode() ERROR!");
				return -1;
			} else {
				Debug.Log("C# Rehamove change_mode(): Changed mode to " + mode.ToString() + ".");
				mode = new_mode;
				return 0;
			}
		}

		public int set_pulse(float current, int pulse_width) {
			if (rehamove == null) {
				Debug.Log("C# Rehamove set_pulse() ERROR! Rehamove object does not exist.");
				return -1;
			}
			int result = rehamovelib.set_pulse_data(rehamove, current, pulse_width);
			if (result != 0) {
				Debug.Log("C# Rehamove set_pulse() ERROR!");
				return -1;
			} else {
				Debug.Log("C# Rehamove set_pulse(): Set pulse current to " + current.ToString() + " and pulse width to " + pulse_width.ToString() + ".");
				return 0;
			}
		}

		public int run(string channel, float period, int total_milliseconds) {
			int chosen_channel = get_channel(channel);
			return run(chosen_channel, period, total_milliseconds);
		}

		public int run(int channel, float period, int total_milliseconds) {
			if (rehamove == null) {
				Debug.Log("C# Rehamove run() ERROR! Rehamove object does not exist.");
				return -1;
			}
			if (mode != MODE_MID_LEVEL) {
				Debug.Log("C# Rehamove run() ERROR! Mode must be set to mid-level.");
				return -1;
			}
			int result = rehamovelib.run(rehamove, channel, period, total_milliseconds);
			if (result != 0) {
				Debug.Log("C# Rehamove run() ERROR!");
				return -1;
			} else {
				Debug.Log("C# Rehamove run(): Completed.");
				return 0;
			}
		}

		public int start(string channel, float period) {
			int chosen_channel = get_channel(channel);
			return start(chosen_channel, period);
		}

		public int start(int channel, float period) {
			if (rehamove == null) {
				Debug.Log("C# Rehamove start() ERROR! Rehamove object does not exist.");
				return -1;
			}
			if (mode != MODE_MID_LEVEL) {
				Debug.Log("C# Rehamove start() ERROR! Mode must be set to mid-level.");
				return -1;
			}
			int result = rehamovelib.midlevel_start(rehamove, channel, period);
			if (result != 0) {
				Debug.Log("C# Rehamove start() ERROR!");
				return -1;
			} else {
				Debug.Log("C# Rehamove start(): Completed.");
				return 0;
			}
		}
		
		public int update() {
			if (rehamove == null) {
				Debug.Log("C# Rehamove update() ERROR! Rehamove object does not exist.");
				return -1;
			}
			if (mode != MODE_MID_LEVEL) {
				Debug.Log("C# Rehamove update() ERROR! Mode must be set to mid-level.");
				return -1;
			}
			int result = rehamovelib.midlevel_update(rehamove);
			if (result != 0) {
				Debug.Log("C# Rehamove update() ERROR!");
				return -1;
			} else {
				Debug.Log("C# Rehamove update(): Completed.");
				return 0;
			}
		}

		public int end() {
			if (rehamove == null) {
				Debug.Log("C# Rehamove end() ERROR! Rehamove object does not exist.");
				return -1;
			}
			if (mode != MODE_MID_LEVEL) {
				Debug.Log("C# Rehamove end() ERROR! Mode must be set to mid-level.");
				return -1;
			}
			int result = rehamovelib.midlevel_end(rehamove);
			if (result != 0) {
				Debug.Log("C# Rehamove end() ERROR!");
				return -1;
			} else {
				Debug.Log("C# Rehamove end(): Completed.");
				return 0;
			}
		}

		public int close() {
			if (rehamove == null) {
				Debug.Log("C# RehamoveDevice close() ERROR! RehamoveDevice object does not exist.");
				return -1;
			}
			int result = rehamovelib.close_port(rehamove);
			if (result != 0) {
				Debug.Log("C# RehamoveDevice close() ERROR!");
				return -1;
			} else {
				Debug.Log("C# RehamoveDevice close() sent.");
				return 0;
			}
		}

		~Rehamove() {
			Debug.Log("I'm being destroyed!");
		}
	}
}

File: builds/python/linux_amd64/rehamove.py
import rehamovelib

class Rehamove:

	current_version = "v1.6"

	channel0 = ['r', 'red']
	channel1 = ['b', 'blue']
	channel2 = ['g1', 'gray1', 'grey1', 'black']
	channel3 = ['g2', 'gray2', 'grey2', 'white']

	MODE_LOW_LEVEL = 0
	MODE_MID_LEVEL = 1

	def __init__(self, port_name):
		self.rehamove = rehamovelib.open_port(port_name)
		self.mode = 0

	def version(self):
		c_version = rehamovelib.get_version()
		print("Rehamove Version: Python-side " + str(Rehamove.current_version) + ", C-side " + str(c_version))
		return Rehamove.current_version

	def get_mode(self):
		result = rehamovelib.get_mode(self.rehamove)
		return result

	def get_current(self):
		result = rehamovelib.get_current(self.rehamove)
		return result

	def get_pulse_width(self):
		result = rehamovelib.get_pulse_width(self.rehamove)
		return result

	def info(self):
		mode = self.get_mode()
		current = self.get_current()
		pulse_width = self.get_pulse_width()

		if mode == Rehamove.MODE_LOW_LEVEL:
			return "Rehamove device in low-level mode. Mid-level pulse is set to {} mA and {} us.".format(current, pulse_width)
		elif mode == Rehamove.MODE_MID_LEVEL:
			return "Rehamove device in mid-level mode. Mid-level pulse is set to {} mA and {} us.".format(current, pulse_width)
		else:
			return "Rehamove info() ERROR!"

	def get_channel(self, channel):
		chosen_channel = channel
		if isinstance(channel, str):
			channel = channel.lower()
			if channel in Rehamove.channel0:
				chosen_channel = 0
			elif channel in Rehamove.channel1:
				chosen_channel = 1
			elif channel in Rehamove.channel2:
				chosen_channel = 2
			elif channel in Rehamove.channel3:
				chosen_channel = 3
			else:
				chosen_channel = 0 # Default
		elif isinstance(channel, int):
			if channel < 0 and channel > 3:
				chosen_channel = 0 # Default
		else:
			chosen_channel = 0
		return chosen_channel

	def pulse(self, channel, current, pulse_width):
		if self.rehamove == None:
			print("python Rehamove pulse() ERROR! Rehamove object does not exist.")
			return -1
		chosen_channel = self.get_channel(channel)
		result = rehamovelib.pulse(self.rehamove, chosen_channel, current, pulse_width)
		if result != 0:
			print("python Rehamove pulse() ERROR!")
			return -1
		else:
			print("python Rehamove pulse() sent.")
			return 0

	def custom_pulse(self, channel, points_array):
		if self.rehamove == None:
			print("python Rehamove custom_pulse() ERROR! Rehamove object does not exist.")
			return -1
		chosen_channel = self.get_channel(channel)
		original_length = len(points_array)
		num_points = len(points_array)
		# Error handling (warning) if too many points.
		if num_points > 16:
			print("python Rehamove custom_pulse() WARNING! Maximum of 16 points allowed, truncating points array.")
			num_points = 16
		
		# Error handling (exception) if malformed points.
		try:
			for i in range(0, num_points):
				current = points_array[i][0]
				pulse_width = points_array[i][1]
		except:
			print("python Rehamove custom_pulse() ERROR! Malformed points array, should be: [ (current0, pulse_width0), (current1, pulse_width1), ... ]")
			return -1

		# Handle if the user supplies less than 16 points: fill up empty points in the array.
		remaining_points = 16 - num_points
		for _ in range(0, remaining_points):
			points_array.append((0.0, 0))

		# Get all of our point data.
		c0, w0 = points_array[0][0], points_array[0][1]
		c1, w1 = points_array[1][0], points_array[1][1]
		c2, w2 = points_array[2][0], points_array[2][1]
		c3, w3 = points_array[3][0], points_array[3][1]
		c4, w4 = points_array[4][0], points_array[4][1]
		c5, w5 = points_array[5][0], points_array[5][1]
		c6, w6 = points_array[6][0], points_array[6][1]
		c7, w7 = points_array[7][0], points_array[7][1]
		c8, w8 = points_array[8][0], points_array[8][1]
		c9, w9 = points_array[9][0], points_array[9][1]
		c10, w10 = points_array[10][0], points_array[10][1]
		c11, w11 = points_array[11][0], points_array[11][1]
		c12, w12 = points_array[12][0], points_array[12][1]
		c13, w13 = points_array[13][0], points_array[13][1]
		c14, w14 = points_array[14][0], points_array[14][1]
		c15, w15 = points_array[15][0], points_array[15][1]

		result = rehamovelib.custom_pulse(self.rehamove, chosen_channel, original_length, c0, w0, c1, w1, c2, w2, c3, w3, c4, w4, c5, w5, c6, w6, c7, w7, c8, w8, c9, w9, c10, w10, c11, w11, c12, w12, c13, w13, c14, w14, c15, w15)
		if result != 0:
			print("python Rehamove custom_pulse() ERROR!")
			return -1
		else:
			print("python Rehamove custom_pulse() sent.")
			return 0

	def battery(self):
		if self.rehamove == None:
			print("python Rehamove ERROR! Rehamove object does not exist.")
			return -1
		result = rehamovelib.battery_request(self.rehamove)
		if result != 0:
			print("python Rehamove battery() ERROR!")
			return -1
		else:
			battery_level = rehamovelib.get_battery(self.rehamove)
			print("python Rehamove battery(): " + str(battery_level) + "%")
			return battery_level

	def change_mode(self, mode):
		if self.rehamove == None:
			print("python Rehamove change_mode() ERROR! Rehamove object does not exist.")
			return -1
		result = rehamovelib.change_mode(self.rehamove, mode)
		if result != 0:
			print("python Rehamove change_mode() ERROR!")
			return -1
		else:
			print("python Rehamove change_mode(): Changed mode to " + str(mode) + ".")
			self.mode = mode
			return 0

	def set_pulse(self, current, pulse_width):
		if self.rehamove == None:
			print("python Rehamove set_pulse() ERROR! Rehamove object does not exist.")
			return -1
		result = rehamovelib.set_pulse_data(self.rehamove, current, pulse_width)
		if result != 0:
			print("python Rehamove set_pulse() ERROR!")
			return -1
		else:
			print("python Rehamove set_pulse(): Set pulse current to " + str(current) + " and pulse width to " + str(pulse_width) + ".")
			return 0

	def run(self, channel, period, total_milliseconds):
		if self.rehamove == None:
			print("python Rehamove run() ERROR! Rehamove object does not exist.")
			return -1
		if self.mode != Rehamove.MODE_MID_LEVEL:
			print("python Rehamove run() ERROR! Mode must be set to mid-level.")
			return -1
		chosen_channel = self.get_channel(channel)
		result = rehamovelib.run(self.rehamove, chosen_channel, period, total_milliseconds)
		if result != 0:
			print("python Rehamove run() ERROR!")
			return -1
		else:
			print("python Rehamove run(): Completed.")
			return 0

	def start(self, channel, period):
		if self.rehamove == None:
			print("python Rehamove start() ERROR! Rehamove object does not exist.")
			return -1
		if self.mode != Rehamove.MODE_MID_LEVEL:
			print("python Rehamove start() ERROR! Mode must be set to mid-level.")
			return -1
		chosen_channel = self.get_channel(channel)
		result = rehamovelib.midlevel_start(self.rehamove, chosen_channel, period)
		if result != 0:
			print("python Rehamove start() ERROR!")
			return -1
		else:
			print("python Rehamove start(): Completed.")
			return 0

	def update(self):
		if self.rehamove == None:
			print("python Rehamove update() ERROR! Rehamove object does not exist.")
			return -1
		if self.mode != Rehamove.MODE_MID_LEVEL:
			print("python Rehamove update() ERROR! Mode must be set to mid-level.")
			return -1
		result = rehamovelib.midlevel_update(self.rehamove)
		if result != 0:
			print("python Rehamove update() ERROR!")
			return -1
		else:
			print("python Rehamove update(): Completed.")
			return 0

	def end(self):
		if self.rehamove == None:
			print("python Rehamove end() ERROR! Rehamove object does not exist.")
			return -1
		if self.mode != Rehamove.MODE_MID_LEVEL:
			print("python Rehamove end() ERROR! Mode must be set to mid-level.")
			return -1
		result = rehamovelib.midlevel_end(self.rehamove)
		if result != 0:
			print("python Rehamove end() ERROR!")
			return -1
		else:
			print("python Rehamove end(): Completed.")
			return 0

	def __del__(self):
		# Only close the port if we have a Rehamove object to close
		if self.rehamove != None:
			result = rehamovelib.close_port(self.rehamove)
			if result != 0:
				print("python Rehamove close_port() ERROR!")

